### DISCUSSION TITLE
Ide

### DISCUSSION THREAD

**Discussion by ddkwork** [vlang/v#14685] [General]
Which ide your used?

**Comment by changrui** [comment]
` VS Code` or `sublime text`.

**Reply by Wajinn** [reply]
You can find a list on awesome-v (https://github.com/vlang/awesome-v):

https://github.com/vlang/awesome-v#syntax-highlighting
https://github.com/vlang/awesome-v#editor-plugins
https://github.com/vlang/notepad-plus-plus (though its getting a bit old)

**Comment by JalonSolov** [comment]
There is also a simple IDE, written in pure V:  https://github.com/IsaiahPatton/Vide/

**Reply by ddkwork** [reply]
> There is also a simple IDE, written in pure V: https://github.com/IsaiahPatton/Vide/

That don't work well for me. Can you creat a github workspace for debugger v  code, Thank you.

**Comment by nkev** [comment]
For anyone looking more recently, there is a new comprehensive plugin with full debugging for Jetbrains IDEs at https://plugins.jetbrains.com/plugin/20287-vlang/docs/quick-start-guide.html

**Reply by i582** [reply]
See also website, where its main features are described:
https://intellij-v.github.io

**Reply by Wajinn** [reply]
> See also website, where its main features are described: https://intellij-v.github.io

This looks great.  But was surprised by some of the review comments that had been made.

**Reply by i582** [reply]
> > See also website, where its main features are described: https://intellij-v.github.io
> 
> This looks great. But was surprised by some of the review comments that had been made.

Thanks for rating the plugin! The Internet remembers everything, and unfortunately the echoes of the events of the V release will sometimes be heard for a long time.
Trust is earned the hard way, but it is lost in an instant and it is very difficult to regain it, which is why we should not lie to users.

**Reply by Wajinn** [reply]
> The Internet remembers everything, and unfortunately the echoes of the events of the V release will sometimes be heard for a long time. Trust is earned the hard way, but it is lost in an instant and it is very difficult to regain it, which is why we should not lie to users.

For random unknowns to make inappropriate remarks trashing others is bad enough, but to echo that to potential users, is very shocking.  It was completely unnecessary for them to write that, because it had nothing to do with the product presented.  We should ask ourselves why they came there to do that?  

What lie?  There are 2 sides of a story.  Why take their side, and not give the benefit of the doubt to the person who is still working hard to achieve their vision and dream?  To add to that, missing a deadline or underestimating the difficulty of something, is not the same as lying.  In fact, there is a strong argument and evidence to say that the other side is lying or distorting to paint a negative narrative.

I've researched and read many perspectives about what has happened, and never came away with the language creator was lying to anybody.  Those who were attempting to say something was going on, are often vague, unspecific, and unclear about what.  But at the same time are trying desperately to persuade others to not use something.  

Upon studying it, often those doing the smearing have very clear agendas or conflicting interests.  Among them, having asked people to use their programming language instead or to donate money to themselves.  That people would overlook the opposing sides reasons to smear or go on smear campaigns is very odd.

**Reply by i582** [reply]
> For random unknowns to make inappropriate remarks trashing others is bad enough, but to echo that to potential users, is very shocking. It was completely unnecessary for them to write that, because it had nothing to do with the product presented. We should ask ourselves why they came there to do that?

I agree, it’s not worth writing such a comment, but it makes no sense to blame someone over whom we have no influence. The main thing we can do is learn from this situation and not repeat the mistakes that led to such comments.

**Reply by i582** [reply]
> Also, what lie? There are 2 sides of a story. Why take their side, and not give the benefit of the doubt to the person who is still working hard to achieve their vision and dream? I've researched and read many perspectives about what has happened, and never came away with the language creator was lying to anybody. Those who were attempting to say something was going on, are often vague, unspecific, and unclear about what. But at the same time are trying desperately to persuade others to not use something.
> 
> Upon studying it, often those doing the smearing have very clear agendas or conflicting interests. Among them, having asked people to use their programming language instead or to donate money to themselves. That people would overlook the opposing sides reasons to smear or go on a smear campaigns is very odd.

See https://andrewkelley.me/post/why-donating-to-musl-libc-project.html

The vlang.io back in 2019 contained a list of features that were described as ready-made (point 3 from article). After the release, when it became clear that not everything was ready, "WIP" labels appeared on the site (item 4-5). That is, all the time before the actual release, the site promised one set of features, and after it turned out that some of the features were missing, the site began to claim that they were WIP.
How would you describe it?

**Reply by Wajinn** [reply]
> I agree, it’s not worth writing such a comment, but it makes no sense to blame someone over whom we have no influence. The main thing we can do is learn from this situation and not repeat the mistakes that led to such comments.

On this we are in agreement.  

Maybe, you would consider to modify the reply to the review.  That some unknown or crazy person gives a weird review, many won't care so much or see it for what they are trying to do.  Definitely you don't have to, but I think people will consider or react to your comment much more strongly.  People love what you created, so your comments have actual weight.

**Reply by i582** [reply]
> Maybe, you would consider to modify the reply to the review.

I think it's a good idea, I wrote it under the influence of resentment that because of the actions of other people my work is suffering and I do not like it.

**Reply by Wajinn** [reply]
First, Andrew and Ginger are **direct competitors**.  Anything they are saying and how they are framing an opposing language, must be taken with a huge grain of salt.  They were on the attack (as it can't be called anything else), even before the language's open-source release.  As I mentioned, they made direct pleas (Ginger directly and for them both) for people to instead use their programming languages and give them money.  There are those that would say that Andrew appeared jealous at the amount of support that V was getting at the time.  The progress and popularity of his language had stalled for a bit and around that time as well.

> How would you describe it?

I would describe it as a difference in opinion and interests.  One side will frame it one way, because they think it will benefit them.  Another side, a different way, as a matter of adjusting expectations and where they are at on delivery.

Vlang was a just released alpha in June 2019.  To think any language would be highly polished upon release is a distortion of reality.  With any programming language, there will be a list of features and claims that somebody can dispute.  That's not a lie, but a difference in opinion.  The vote on the trust or validity, is done by the supporters.  Many of V's backers and supporters continued their support, because they believed and trusted in it's creator.  It's why we are here today.

**Reply by Wajinn** [reply]
> I think it's a good idea, I wrote it under the influence of resentment that because of the actions of other people my work is suffering and I do not like it.

You will have supporters.  People are viewing your work favorably.  Don't allow the negativity or competing interests of others to hinder your enthusiasm.  You can win in the end.

**Reply by i582** [reply]
> I would describe it as a difference in opinion and interests. One side will frame it one way, because they think it will benefit them. Another side, a different way, as a matter of adjusting expectations and where they are at on delivery.

If you buy a product in a store with one description and get a different one, do you also think that this is a "difference in opinion and interests"? The project collected donations describing some features for ready ones, although they were not, this is a direct deceit of expectations and a decrease in donations after the release clearly shows that people realized that they were deceived, since they received far from what was described.

**Reply by i582** [reply]
Donations are the main problem, as I understand this situation, because it's like a kickstarter, where people raise money for cool projects.
If you end up releasing something completely different from what you promised, it's a deceit and it doesn't matter who wrote about it, even if this person creates something similar to yours.

If the project was just doing something without donations, then everything you write would be applicable and I would agree. Probably there would not be so many haters, since this would be one of the millions of projects that promise one thing and give another.

**Reply by Wajinn** [reply]
A programming language is a bit different from a regular consumer product.  It's not going to be released as a finished product. Supporters familiar with the language development process would know this, and the majority maintained their support of V.  

It's my opinion that direct competitors and the opposition are going to frame a situation to their benefit and view to gain customers and users at the other's expense.  That a few of V's supporters were "knocked off" and possibly persuaded by their rhetoric, is their intention.  However, of the few that were knocked off, V has gained huge numbers of new supporters.  What they are accusing or insinuating about V, others have insinuated about their programming language (just have to study up on them enough), and that they failed to deliver on features and claims.  This is common for programming languages in general.  People can also be fickle, and jump ship, for nearly anything.  Some people will assume guilt or lose faith, just by rumor or hearing of an accusation.  It doesn't mean everybody else or the majority will.

V has maintained a tremendous amount of support. I've read numerous comments of direct financial supporters for V from that time to the present.  They are still enthusiastic and supportive of V's progress and destination.  They understood it was an alpha release of a programming language, and that no programming languages is a finished product upon release.  Thus, we are here today.  V, in comparison, has progressed a lot faster than many other languages.

> Probably there would not be so many haters...

There would be haters, regardless. V was both unexpected and disruptive, from the perspective of its competitors. It seemingly came out of nowhere, got supporters, continued to progress, and has done well. What they see as V's success, can be wrongfully looked at as what they should have got or they could have got more if it didn't exist. These types of negative views are going to breed jealous, contempt, and hate.

We are way past 2019. Seems way more beneficial to move forward. Celebrate the positive and progress.

**Reply by medvednikov** [reply]
What reviews are you talking about? All reviews here are positive: https://plugins.jetbrains.com/plugin/20287-vlang/reviews

**Reply by i582** [reply]
> What reviews are you talking about? All reviews here are positive: https://plugins.jetbrains.com/plugin/20287-vlang/reviews

Hmm, apparently the author deleted it, because I did nothing.

**Reply by medvednikov** [reply]
@i582 there was no lying, only lying was from the attacks on V (like "no way V is written in V", "V is not fast", "V compiler depends on OpenGL" etc). I'm tired of all these lies from 2019 still coming up and I'll do a big debunking blog post soon.

Your example with buying a product in a store is not correct. This is not a product you buy, it's free and open source. People are free to support it if they want, and none of the supporters felt they were scammed.

**Reply by medvednikov** [reply]
> decrease in donations after the release clearly shows that people realized that they were deceived, since they received far from what was described.

No it does not. People were paying $30 for the pre-release build of V. Once it came out, some no longer wanted to pay. That resulted in that whopping $100/month drop.

**Reply by i582** [reply]
> @i582 there was no lying, only lying was from the attacks on V (like "no way V is written in V", "V is not fast", "V compiler depends on OpenGL" etc). I'm tired of all these lies from 2019 still coming up and I'll do a big debunking blog post soon.
> 
> Your example with buying a product in a store is not correct. This is not a product you buy, it's free and open source. People are free to support it if they want, and none of the supporters felt they were scammed.

You and I have already discussed this in private messages and you and I have different opinions on this matter, I don’t think it’s worth discussing it again. Above, I have provided some of the data on which I drew conclusions, so that any reader can find out for himself what he thinks about it.

**Reply by i582** [reply]
> No it does not. People were paying $30 for the pre-release build of V. Once it came out, some no longer wanted to pay. That resulted in that whopping $100/month drop.

That is, people also paid for access to a product that was described in one way, but turned out to be with some unfinished features that were not indicated as such?

**Reply by medvednikov** [reply]
Well it's good that it was removed. Your plugin has so many good positive comments, and you let one idiot get to you.

No matter how good a project is, if it's popular, there will be a minority of people shitting on it no matter what.

**Reply by medvednikov** [reply]
> That is, people also paid for access to a product that was described in one way, but turned out to be with some unfinished features that were not indicated as such?

And yet none of them had a problem with that and continued the support because they understood it was a WIP project.

By the way you keep referring to this REPL delayed by 1 week, like it's a big deal and it scammed paying customers.

You do know that that pre-built V had REPL, and it was only delayed by 1 week for the open source release, right? :)

**Reply by medvednikov** [reply]
> You and I have already discussed this in private messages 

We did, that's why I'm surprised you bring up that donation thing, when I already explained that before the open source release, V could be bought for $30 on Patreon, so of course Patreon donations would go down after the release.

**Reply by i582** [reply]
> > That is, people also paid for access to a product that was described in one way, but turned out to be with some unfinished features that were not indicated as such?
> 
> And yet none of them had a problem with that and continued the support because they understood it was a WIP project.
> 
> By the way you keep referring to this REPL delayed by 1 week, like it's a big deal and it scammed paying customers.
> 
> You do know that that pre-built V had REPL, and it was only delayed by 1 week for the open source release, right? :)

Compare these two versions. Seems like WIP is added for more than just a REPL, doesn't it?

Before release: https://web.archive.org/web/20190619233707/https://vlang.io/#
Two day after: https://web.archive.org/web/20190624052507/https://vlang.io/ (11 new WIP)

**Reply by i582** [reply]
> > You and I have already discussed this in private messages
> 
> We did, that's why I'm surprised you bring up that donation thing, when I already explained that before the open source release, V could be bought for $30 on Patreon, so of course Patreon donations would go down after the release.

I want to admit, yes, apparently here my conclusions are not very accurate and I forgot part of our communication. Well, that's just one small part that has little influence on what I'm saying.

**Reply by medvednikov** [reply]
Yes, not having WIPs was a mistake, and I added missing WIPs (the zero dependencies was later changed to zero library dependencies, it's still an easy bootstrap with `cc v.c`), and with time these features were implemented, and all WIPs were removed, and all points on the website were very carefully checked, so that all of them are true. I got lots of help and feedback with that, including from you.

Everyone makes mistakes, I fixed it, it was >3 years ago. V is much bigger now and is the result of work of 600+ contributors, and they don't deserve these accusations for a mistake I made.

**Reply by Wajinn** [reply]
> What reviews are you talking about? 
 
> Hmm, apparently the author deleted it, because I did nothing.

That kind of slanderous review had nothing to do with the product.  Apparently, it was in violation of [JetBrains policies](https://plugins.jetbrains.com/docs/marketplace/reviews-policy.html), so got removed when they got around to looking at it.  Which is great.  It was a very foul thing to do, which shows what some people will resort to and what's going on.

**Reply by Wajinn** [reply]
> Compare these two versions. 

Don't have a problem with it (and apparently many others).  I'm of the opinion that it is unfortunate to allow competitors (who mean to cause confusion and pull away users) to frame one's point of view, which might be happening here.  The website is being viewed from their context and narrative.  That is, any changes, updates, or clarifications of status and state of product, is somehow malicious or intentional deception (when they know it's not).  

When they crafted such a negative narrative against V, they knew what they were doing was unfair and deceptive, because their languages were in such unfinished and unpolished states (WIP) when released and their websites have gone through iterations of changes related to features and claims.  They were also taking advantage that their languages were released many years before, so of course at that moment (in 2019), they were ahead.  But a much more fair comparison is the state of their languages and websites, upon first public release.  And it's worth mentioning, how quickly V caught up or passed various languages in terms of progress and popularity (which can cause resentment).

An alpha release of a programming language is just about impossible to get perfect and not be WIP.  As many webmasters and those in marketing know, changes to the website are also done to be informative, make corrections, and reflect status.

**Reply by Wajinn** [reply]
>  decrease in donations after the release clearly shows that people realized that they were deceived, since they received far from what was described.

> No it does not. People were paying $30 for the pre-release build of V. Once it came out, some no longer wanted to pay. That resulted in that whopping $100/month drop.


After re-reading the previous posts and more history, I'm going to add a bit more (at the risk of overkill), because I think there is a disconnect on a particular point and which comes from confusion and distortions generated by Andrew's negative framing (back in 2019).  I'm only mentioning this, because an old link of his was posted in this thread and used as if proof of something.

Andrew (who is a direct competitor), appeared to make an effort to break off supporters and/or draw additional funding towards himself and projects benefiting him, at the expense of V.  And did so numerous times, on different occasions.  People might not be aware of the history or context, but if dug into enough, it's pretty clear and obvious.  Andrew (and others in a similar position) should not be used as an honest objective party (when clearly not) or as any kind of valid source.  

There were people paying for pre-releases of V, which were not available prior to the open source general public release in June of 2019.  This is often called early access.  This is a very legitimate and valid way of promoting products, showing support, and helping to fund work on it.  In this case, the end goal is the release of a programming language.  An alpha release of a programming language, almost by definition, is a WIP.  Nobody in their right mind would have the expectation of a finished language on public release.  Rather, as paying supporters, they would be able to see and use the language before it goes public.  Paying for early access, does not mean expecting and paying for a specific feature set that must be or was guaranteed to be finished by a set date.

That is a key point to be clear on, because among Andrew's arguments, was accusing the V project of fooling people into donating to it (in 2019) because it was a closed source project behind a paywall.  That's flat wrong or purposeful distortion.  Those people were getting early access to the language, as paying supporters, so they can use it or see for themselves how things were coming along.  

When the V language was publicly released, those with the motivation of paying specifically for early access, now had access.  So they may have decided to no longer donate.  That's much different from Andrew's misleading narrative of there being a "backlash" because they were "deceived".  How were they "deceived"?  They had early access to use the language, knew more about what was going on than anybody else, and had access to its creator.  And the majority of V supporters continued donating, to further language development.

This should also be seen in the context and the narrative of deception promoted by competitors, where V was said to be vaporware or a scam.  As in V didn't actually exist, and that people were supporting and donating for nothing, thus they should support and give their money to them (competitors).  When it was clear that V really did exist and was publicly released, the narratives quickly shifted to finding anything about the website that could be questioned or features being WIP of a newly released alpha.
